src/sdk/pynni/nni/compression/torch/builtin_pruners.py

```diff
@@ -102,7 +182,168 @@ def compute_target_sparsity(self, config):
+
+    Hao Li, Asim Kadav, Igor Durdanovic, Hanan Samet and Hans Peter Graf,
+    "PRUNING FILTERS FOR EFFICIENT CONVNETS", 2017 ICLR
+    https://arxiv.org/abs/1608.08710
+
+    Attributes
+    ----------
+    mask_dict : dict
+        Dictionary for saving masks.
+
+    """
+
+    def __init__(self, model, config_list):
+        """
+        Initiate `FilterPruner` from `config_list`
+        config_list: supported keys:
+            - sparsity
+
+        Parameters
+        ----------
+        model : torch.nn.module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+
+        """
+
+        super().__init__(model, config_list)
+        self.if_init_list = {}
+
+    def calc_mask(self, layer, config):
+        """
+        Calculate the mask of given layer
+
+        Parameters
+        ----------
+        layer : LayerInfo
+            the layer to instrument the compression operation
+        config : dict
+            layer's pruning config
+
+        Returns
+        -------
+        torch.Tensor
+            mask of the layer's weight
+
+        """
+
+        weight = layer.module.weight.data
+        op_name = layer.name
+        op_type = layer.type
+        assert op_type == 'Conv2d', 'FilterPruner only supports 2d convolution layer pruning'
+        if self.if_init_list.get(op_name, True):
+            kernels = weight.shape[0]
+            w_abs = weight.abs()
+            k = int(kernels * config['sparsity'])
+            if k == 0:
+                return torch.ones(weight.shape).type_as(weight)
+            w_abs_structured = w_abs.view(kernels, -1).sum(dim=1)
+            threshold = torch.topk(w_abs_structured.view(-1), k, largest=False).values.max()
+            mask = torch.gt(w_abs_structured, threshold)[:, None, None, None].expand_as(weight).type_as(weight)
+            self.mask_dict.update({op_name: mask})
+            self.if_init_list.update({op_name: False})
+        else:
+            mask = self.mask_dict[op_name]
+        return mask
+
+
+class SlimPruner(Pruner):
+    """
+    A structured pruning algorithm that prunes channels by pruning the weights of BN layers.
+    Zhuang Liu, Jianguo Li, Zhiqiang Shen, Gao Huang, Shoumeng Yan and Changshui Zhang
+    "Learning Efficient Convolutional Networks through Network Slimming", 2017 ICCV
+    https://arxiv.org/pdf/1708.06519.pdf
+
+    Attributes
+    ----------
+    mask_dict : dict
+        Dictionary for saving masks.
+
+    """
+
+    def __init__(self, model, config_list):
+        """
+        Initiate `FilterPruner` from `config_list`
+        config_list: supported keys:
+            - sparsity
+
+        Parameters
+        ----------
+        config_list : list
+            List of pruning configs
+
+        """
+
+        super().__init__(model, config_list)
+        self.if_init_list = {}
+        weight_list = []
+        config = self.config_list[0]
```

Comment:
please do not directly use `self.config_list`, use `config_list` instead.

If you only use the first config, then warning if `self.config_list` has more than one elements.